src/faiss.cc

```diff
@@ -335,11 +426,143 @@ class IndexFlatL2 : public Napi::ObjectWrap<IndexFlatL2>
     delete[] xb;
     return Napi::Number::New(info.Env(), num);
   }
+
+  Napi::Value toBuffer(const Napi::CallbackInfo &info)
+  {
+    Napi::Env env = info.Env();
+
+    if (info.Length() != 0)
+    {
+      Napi::Error::New(env, "Expected 0 arguments, but got " + std::to_string(info.Length()) + ".")
+          .ThrowAsJavaScriptException();
+      return env.Undefined();
+    }
+
+    faiss::VectorIOWriter *writer = new faiss::VectorIOWriter();
+
+    faiss::write_index(index_.get(), writer);
+
+    // return buffer from IOWriter
+    return Napi::Buffer<uint8_t>::Copy(env, writer->data.data(), writer->data.size());
+  }
+
+protected:
+  std::unique_ptr<faiss::Index> index_;
+  inline static Napi::FunctionReference *constructor;
+};
+
+// faiss::Index is abstract so IndexFlatL2 is used as fallback
+class Index : public IndexBase<Index, faiss::IndexFlatL2>
+{
+public:
+  using IndexBase::IndexBase;
+
+  static constexpr const char *CLASS_NAME = "Index";
+
+  static Napi::Object Init(Napi::Env env, Napi::Object exports)
+  {
+    // clang-format off
+    auto func = DefineClass(env, CLASS_NAME, {
+      InstanceMethod("ntotal", &Index::ntotal),
+      InstanceMethod("getDimension", &Index::getDimension),
+      InstanceMethod("isTrained", &Index::isTrained),
+      InstanceMethod("add", &Index::add),
+      InstanceMethod("train", &Index::train),
+      InstanceMethod("search", &Index::search),
+      InstanceMethod("write", &Index::write),
+      InstanceMethod("mergeFrom", &Index::mergeFrom),
+      InstanceMethod("removeIds", &Index::removeIds),
+      InstanceMethod("toBuffer", &Index::toBuffer),
+      StaticMethod("read", &Index::read),
+      StaticMethod("fromBuffer", &Index::fromBuffer),
+      StaticMethod("fromFactory", &Index::fromFactory),
+    });
+    // clang-format on
+
+    constructor = new Napi::FunctionReference();
```

Comment:
Not as clean as constructing within Base, but that was causing stability issues. I don't fully understand the difference, but this works around whatever was going on.

Reply:
What kind of stability issues is it and how to reproduce it?

Reply:
Bun runtime (https://bun.sh/) segfaults via `bun test`.

Reply:
bun test still failed in `test/IndexFlatL2.test.js`
<img width="543" alt="image" src="https://github.com/ewfian/faiss-node/assets/12423122/fdc63891-66f7-43e0-b180-cbed0dd0bd85">

after remove this case, it works
<img width="806" alt="image" src="https://github.com/ewfian/faiss-node/assets/12423122/72af27ff-48e7-4a86-b7d3-2f1ca140c585">

my test is that the following code will never be reached.
<img width="706" alt="image" src="https://github.com/ewfian/faiss-node/assets/12423122/7087672f-952e-4fce-b2d3-830408d2af38">


and the Error was thrown by napi internally
<img width="684" alt="image" src="https://github.com/ewfian/faiss-node/assets/12423122/69daea99-e21d-4cf1-8556-a841345acc55">

Reply:
Yeah I'm fine merging with that issue for now since I created the issue on the Bun side. https://github.com/oven-sh/bun/issues/4526

Reply:
thank you, i will remove this test case and the following useless code

![](https://user-images.githubusercontent.com/12423122/268990930-7087672f-952e-4fce-b2d3-830408d2af38.png)

Reply:
seems that this case still failed on bun test

<img width="929" alt="image" src="https://github.com/ewfian/faiss-node/assets/12423122/a60fcd80-2bf8-4dc4-b079-486f36d78766">

Reply:
Yes all tests which rely on throwing won't work in Bun until the issue I linked is resolved. That's fine, and not this packages problem.